src/mesh/curved_mesh.jl

```diff
@@ -0,0 +1,170 @@
+"""
+    CurvedMesh{NDIMS, RealT<:Real} <: AbstractMesh{NDIMS}
+
+A structured curved mesh.
+
+Different numbers of cells per dimension are possible and arbitrary functions 
+can be used as domain faces.
+
+Note: This is an experimental feature and may be changed in future releases without notice.
+"""
+mutable struct CurvedMesh{NDIMS, RealT<:Real} <: AbstractMesh{NDIMS}
+  cells_per_dimension::NTuple{NDIMS, Int}
+  faces::Tuple
+  faces_as_string::Vector{String}
+  current_filename::String
+  unsaved_changes::Bool
+end
+
+
+"""
+    CurvedMesh(cells_per_dimension, faces, RealT)
+
+Create a CurvedMesh of the given size and shape that uses `RealT` as coordinate type.
+
+# Arguments
+- `cells_per_dimension::NTupleE{NDIMS, Int}`: the number of cells in each dimension.
+- `faces::NTuple{2*NDIMS, Function}`: a tuple of `2 * NDIMS` functions that describe the faces of the domain.
+                                      Each function must take `NDIMS-1` arguments.
+                                      `faces[1]` describes the face onto which the face in negative x-direction 
+                                      of the unit hypercube is mapped. The face in positive x-direction of
+                                      the unit hypercube will be mapped onto the face described by `faces[2]`.
+                                      `faces[3:4]` describe the faces in positive and negative y-direction respectively 
+                                      (in 2D and 3D).
+                                      `faces[5:6]` describe the faces in positive and negative z-direction respectively
+                                      (in 3D).
+- `RealT::Type`: The type that should be used for coordinates.
+- `faces_as_string::Vector{String}`: a vector which contains the string of the function definition of each face.
+                                     If `CodeTracking` can't find the function definition, it can be passed directly here.
+"""
+function CurvedMesh(cells_per_dimension, faces, RealT::Type; unsaved_changes=true, faces_as_string=faces2string(faces))
+  NDIMS = length(cells_per_dimension)
+
+  return CurvedMesh{NDIMS, RealT}(cells_per_dimension, faces, faces_as_string, "", unsaved_changes)
+end
+
+
+"""
+    CurvedMesh(cells_per_dimension, coordinates_min, coordinates_max)
+
+Create a CurvedMesh that represents a uncurved structured mesh with a rectangular domain.
+
+# Arguments
+- `cells_per_dimension::NTuple{NDIMS, Int}`: the number of cells in each dimension.
+- `coordinates_min::NTuple{NDIMS, RealT}`: coordinate of the corner in the negative direction of each dimension.
+- `coordinates_max::NTuple{NDIMS, RealT}`: coordinate of the corner in the positive direction of each dimension.
```

Comment:
That sounds reasonable.

Reply:
Does this apply to the discussion above about `cells_per_dimension` too? So we make this a `SVector` too?

Reply:
Ah wait, I remember why we used `NTuple` here. The tuple is passed to `coordinates2faces` and the only way for us to dispatch by dimension is to use a tuple. We would have to use if-else otherwise.

Reply:
These are two orthogonal questions. You can dispatch on the dimension of an `SVector`:
```julia
julia> using StaticArrays

julia> foo(::SVector{1}) = "one"
foo (generic function with 1 method)

julia> foo(::SVector{2}) = "two"
foo (generic function with 2 methods)

julia> foo(SVector(1.0))
"one"

julia> foo(SVector(1.0, 2.0))
"two"
```
The other question is whether it's useful to have an `SVector`. Looking at [KISS](https://en.wikipedia.org/wiki/KISS_principle), I prefer to use the most simple type. Thus, if we need vector space operations (addition, multiplication by scalars), an `SVector` is fine. If we do not need these, why shouldn't we just use a tuple?

Reply:
Yes, using tuples is fine by me (see the other thread on SVector vs Tuple). In any case, I think it would be nice to write the interface in a way that lets the user use *at least* the two reasonable variants of `Tuple` and `Vector` to store coordinate data (and possibly also `SVector`).

Reply:
Should we just change `coordinates2faces` to
```julia
function coordinates2faces(coordinates_min::Union{NTuple{1}, AbstractVector{1}}, coordinates_max::Union{NTuple{1}, AbstractVector{1}})
```
to allow for both vectors and tuples?
We only access it by `coordinates_min[index]` in this function.

Reply:
And we could add a
```julia
cells_per_dimension_ = Tuple(cells_per_dimension)
```
to allow for vectors and tuples here as well.

Reply:
No, since in AbstractVector there is no type information on the length, you cannot dispatch on it (the example you showed would not compile, since it essentially says the element type of the abstract vector should be `1`).

Reply:
How about calling `coordinates2nodes` with `Tuple(coordinates_min)` then? This should allow for any kind of vector or tuple, right?

Reply:
The comment above was in answer to your question "Should we just change...". 
Yes, using `Tuple(...)` sounds like a reasonable idea, since it works with tuples, vectors, and svectors.